docs/common/craft-parts/how-to/customize-the-build-with-craftctl.rst

```diff
@@ -0,0 +1,167 @@
+
+Replace ``<step>`` with the name of a step in the lifecycle.
+
+
+Override a step
+---------------
+
+Craftctl comes in most handy when when you need to tweak files or settings before or
+after executing a step.
+
+To specify where the regular step behavior occurs inside the scriplet, call it with the
+special ``craftctl default`` command. Using that command as an anchor point, alter the build files or environment with other commands before or after it.
+
+.. admonition:: Don't forget to execute the step
+    :class: caution
+
+    If you don't call ``craftctl default`` inside a scriptlet, the step is skipped,
+    and the part doesn't finish building.
+
+Let's say you want to modify a file copied from a part's source. You would add an
+override for the pull step with the ``override-pull`` key, run the default step with
+craftctl, then change the copied source file.
+
+The following example does just that. The icon path inside the part's desktop file in
+the is tweaked after downloading it, but before |app| builds it.
+
+.. code-block:: yaml
+
+    parts:
+      gnome-text-editor:
+        source: https://gitlab.gnome.org/GNOME/gnome-text-editor
+        override-pull: |
+          craftctl default
+          sed -i.bak -e 's|Icon=@app_id@$|Icon=snap.gnome-text-editor.icon|g' data/org.gnome.TextEditor.desktop
+
+
+Override project variables
+--------------------------
+
+With craftctl, you can dynamically set project variables specific to |app| -- distinct
+from environment variables -- within an override scriptlet. Set values with:
+
+.. code-block:: bash
+
+    craftctl set <key>=<value>
+
+You can also retrieve the current value of a project key, with:
+
+.. code-block:: bash
+
+    craftctl get <key>
+
+.. TODO: In Snapcraft, the ``version`` and ``grade`` keys are supported.
+
+For example, if you want to set the ``version`` key of |artifact-indefinite| to the
+latest Git tag in a part's source, declare this override:
+
+.. code-block:: yaml
+    :caption: Project file
+
+    adopt-info: gnome-text-editor
+    parts:
+      gnome-text-editor:
+        source: https://gitlab.gnome.org/GNOME/gnome-text-editor
+        override-pull: |
+          craftctl default
+          craftctl set version=$(git describe --tags --abbrev=10)
+
+.. important:
+
+    .. To incorporate retrieved metadata correctly, ensure that ``adopt-info`` points to
+    the part that runs ``craftctl set``. Without this, snap metadata won't update.
+
+For another example, let's say |artifact-indefinite| has a manually-set version number
+in the project file. To append the latest Git commit hash to the version, declare this
+override:
+
+.. code-block:: yaml
+    :caption: Project file
+
+    adopt-info: gnome-text-editor
+    parts:
+      gnome-text-editor:
+        override-stage: |
+          craftctl default
+          craftctl set version="$(craftctl get version)-$(git rev-parse --short HEAD)"
+
+
+Expose part variables in apps
+-----------------------------
```

Comment:
This section seems to be interesting only to developers writing new applications that use craft-parts, and (unlike the previous sections) it's not interesting for someone using the *craft application.

Reply:
The plan was to keep it in for Craft Parts' docs but exclude it when imported in our apps' docs. Would that work for you?

Reply:
Yes, that's the correct approach!